p2panda-rs/src/document/document_view.rs

```diff
@@ -1,251 +1,206 @@
-        };
-
-        let fields = operation.fields();
+    /// Get the id of this document.
+    pub fn document_id(&self) -> &Hash {
+        self.id.document_id()
+    }
 
-        if let Some(fields) = fields {
-            for (key, value) in fields.iter() {
-                self.0.insert(key.to_string(), value.to_owned());
-            }
-        }
+    /// Get the document view id.
+    pub fn document_view_id(&self) -> &DocumentViewId {
+        &self.id
+    }
 
-        Ok(())
+    /// Get a single value from this instance by it's key.
+    pub fn get(&self, key: &str) -> Option<&OperationValue> {
+        self.view.get(key)
     }
 
     /// Returns a vector containing the keys of this instance.
     pub fn keys(&self) -> Vec<String> {
-        self.0.clone().into_keys().collect::<Vec<String>>()
+        self.view.clone().into_keys().collect::<Vec<String>>()
     }
 
     /// Returns an iterator of existing instance fields.
     pub fn iter(&self) -> BTreeMapIter<String, OperationValue> {
-        self.0.iter()
+        self.view.iter()
     }
 
     /// Returns the number of fields on this instance.
     pub fn len(&self) -> usize {
-        self.0.len()
+        self.view.len()
     }
 
     /// Returns true if the instance is empty, otherwise false.
     pub fn is_empty(&self) -> bool {
-        self.0.is_empty()
-    }
-}
-
-impl TryFrom<Operation> for DocumentView {
-    type Error = DocumentViewError;
-
-    fn try_from(operation: Operation) -> Result<DocumentView, DocumentViewError> {
-        if !operation.is_create() {
-            return Err(DocumentViewError::NotCreateOperation);
-        };
-
-        let mut instance: DocumentView = DocumentView::new();
-        let fields = operation.fields();
-
-        if let Some(fields) = fields {
-            for (key, value) in fields.iter() {
-                instance.0.insert(key.to_string(), value.to_owned());
-            }
-        }
-
-        Ok(instance)
-    }
-}
-
-impl TryFrom<OperationWithMeta> for DocumentView {
-    type Error = DocumentViewError;
-
-    fn try_from(operation: OperationWithMeta) -> Result<DocumentView, DocumentViewError> {
-        if !operation.is_create() {
-            return Err(DocumentViewError::NotCreateOperation);
-        };
-
-        let mut instance: DocumentView = DocumentView::new();
-        let fields = operation.fields();
-
-        if let Some(fields) = fields {
-            for (key, value) in fields.iter() {
-                instance.0.insert(key.to_string(), value.to_owned());
-            }
-        }
-
-        Ok(instance)
+        self.view.is_empty()
     }
 }
-
-impl From<BTreeMap<String, OperationValue>> for DocumentView {
-    fn from(map: BTreeMap<String, OperationValue>) -> Self {
-        Self(map)
-    }
-}
-
-// @TODO: This currently makes sure the wasm tests work as cddl does not have any wasm support
-// (yet). Remove this with: https://github.com/p2panda/p2panda/issues/99
-#[cfg(not(target_arch = "wasm32"))]
```

Comment:
Very happy this slowly gets untied 🥳

Reply:
Yes, it felt great to delete all of these :1st_place_medal: